src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -0,0 +1,224 @@
+                // for any of the provided objects then we throw an exception. 
+                if (rootEntityDescriptor == null)
+                {
+                    throw Error.InvalidOperation(Strings.Context_EntityNotContained);
+                }
+
+                // If it is a top-level object,
+                // meaning that it was sent as top-level object by the user
+                // then we add it to graph's top-level objects. 
+                if (isRootObject)
+                { 
+                    bulkUpdateGraph.AddTopLevelDescriptor(rootEntityDescriptor);
+                }
+
+                foreach (Descriptor descriptor in descriptors)
+                {
+                    if (descriptor is EntityDescriptor entityDescriptor)
+                    {
+                        if (entityDescriptor.ParentEntity != null && 
+                            entityDescriptor.ParentEntity.Equals(rootEntityDescriptor.Entity) && 
+                            !bulkUpdateGraph.Contains(entityDescriptor))
+                        {
+                            bulkUpdateGraph.AddRelatedDescriptor(rootEntityDescriptor, entityDescriptor);
+                            this.BuildDescriptorGraph(descriptors, false, entityDescriptor.Entity);
+                        }
+                    }
+                    else if (descriptor is LinkDescriptor link)
+                    {
+                        if (link.Source.Equals(rootEntityDescriptor.Entity) && !bulkUpdateGraph.Contains(link))
+                        {
+                            bulkUpdateGraph.AddRelatedDescriptor(rootEntityDescriptor, link);
+                            this.BuildDescriptorGraph(descriptors, false, link.Target);
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Creates an ODataRequestMessage for a bulk update request.
+        /// </summary>
+        /// <returns>Returns an instance of the ODataRequestMessage for the bulk update request.</returns>
+        private ODataRequestMessageWrapper CreateBulkUpdateRequest(BulkUpdateGraph bulkUpdateGraph)
+        {
+            EntityDescriptor baseDescriptor = bulkUpdateGraph.TopLevelDescriptors[0] as EntityDescriptor;
+            Uri requestUri = UriUtil.CreateUri(this.RequestInfo.BaseUriResolver.GetBaseUriWithSlash(), UriUtil.CreateUri(bulkUpdateGraph.EntitySetName, UriKind.Relative));
+            Version requestVersion = Util.ODataVersion401;
+            string httpMethod = this.GetHttpMethod(EntityStates.Modified, ref requestVersion);
+
+            HeaderCollection headers = new HeaderCollection();
+
+            this.RequestInfo.Context.Format.SetRequestContentTypeForEntry(headers);
+
+            ApplyPreferences(headers, httpMethod, this.RequestInfo.AddAndUpdateResponsePreference, ref requestVersion);
+
+            headers.SetHeader("OData-Version", "4.01");

```

Comment:
Hard coding the version string?

Reply:
Yes. We have to use this particular version to write a bulk update request..There is a PR to support v4.01 in OData Client as currently it is not supported.